ql/src/test/org/apache/hadoop/hive/ql/exec/TestGetPartitionAuthWithBatches.java

```diff
@@ -0,0 +1,274 @@
+        verify(spyMSC, times(3)).listPartitionsWithAuthInfoRequest(req.capture());
+        req.getAllValues().forEach(part-> Assert.assertEquals(part.getPartNames().size(),10));
+    }
+
+    /**
+     * Tests the number of times Hive.getAllPartitionsOf calls are executed with total number of
+     * partitions to be added are not exactly divisible by batch size
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testUnevenNumberOfGetPartitionCalls() throws Exception {
+        HiveMetaStoreClient spyMSC = spy(msc);
+        hive.setMSC(spyMSC);
+        // there should be 2 calls to get partitions with batch sizes of 19, 11
+        hive.getAllPartitionsInBatches(hive.getTable(dbName, tableName),19, 2, 0, null, true, "user", new ArrayList<>(Arrays.asList("Grp1", "Grp2")));
+        ArgumentCaptor<GetPartitionsPsWithAuthRequest> req = ArgumentCaptor.forClass(GetPartitionsPsWithAuthRequest.class);
+        // there should be 2 calls to get partitions
+        verify(spyMSC, times(2)).listPartitionsWithAuthInfoRequest(req.capture());
+        // confirm the batch sizes were 19, 11 in the two calls to get partitions
+        List<GetPartitionsPsWithAuthRequest> apds = req.getAllValues();
+        Assert.assertEquals(19, apds.get(0).getPartNames().size());
+        Assert.assertEquals(11, apds.get(1).getPartNames().size());
+    }
+
+    /**
+     * Tests the number of times Hive.getAllPartitionsOf calls are executed with total number of
+     * partitions to is less than batch size
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testSmallNumberOfPartitions() throws Exception {
+        HiveMetaStoreClient spyMSC = spy(msc);
+        hive.setMSC(spyMSC);
+        hive.getAllPartitionsInBatches(hive.getTable(dbName, tableName),100, 2, 0, null, true, "user", new ArrayList<>(Arrays.asList("Grp1", "Grp2")));
+        ArgumentCaptor<GetPartitionsPsWithAuthRequest> req = ArgumentCaptor.forClass(GetPartitionsPsWithAuthRequest.class);
+        // there should be 1 call to get partitions
+        verify(spyMSC, times(1)).listPartitionsWithAuthInfoRequest(req.capture());
+        Assert.assertEquals(30, req.getValue().getPartNames().size());
+    }
+
+    /**
+     * Tests the retries exhausted case when getAllPartitionsOf method call always keep throwing
+     * HiveException. The batch sizes should exponentially decreased based on the decaying factor and
+     * ultimately give up when it reaches 0
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testRetriesExhaustedBatchSize() throws Exception {
+        HiveMetaStoreClient spyMSC = spy(msc);
+        hive.setMSC(spyMSC);
+        doThrow(MetaException.class).when(spyMSC).listPartitionsWithAuthInfoRequest(any());
+        try {
+            hive.getAllPartitionsInBatches(hive.getTable(dbName, tableName), 30, 2, 0, null, true, "user", new ArrayList<>(Arrays.asList("Grp1", "Grp2")));
+        } catch (Exception ignored) {}
```

Comment:
Are you expecting HiveException here? If so, why not to catch that itself than catching generic exception?

Reply:
done